File: lib_python/camera.py
from typing import Any

from .vec import Vec3


class Camera:

    def __init__(
        self,
        camera_position: Vec3 = Vec3(0, 0, 0),
        camera_rotation: Vec3 = Vec3(0, 0, 0),
        camera_focal: float = 70,
        camera_clip_start: float = 0.001,
        camera_clip_end: float = 100,
        locked_camera_movement: bool = False,
        locked_camera_rotation: bool = False,
        camera_width: int = 64,
        camera_height: int = 64,
        camera_pixel_size: float = 0.4,
    ) -> None:

        self.camera_position: Vec3 = camera_position
        self.camera_rotation: Vec3 = camera_rotation
        self.camera_focal: float = camera_focal
        self.camera_clip_start: float = camera_clip_start
        self.camera_clip_end: float = camera_clip_end
        self.locked_camera_movement: bool = locked_camera_movement
        self.locked_camera_rotation: bool = locked_camera_rotation
        self.camera_width: int = camera_width
        self.camera_height: int = camera_height
        self.camera_pixel_size: float = camera_pixel_size

    def export_to_dict(self) -> dict[str, Any]:

        return {
            "camera_position": self.camera_position.export_to_str(),
            "camera_rotation": self.camera_rotation.export_to_str(),
            "camera_focal": self.camera_focal,
            "camera_clip_start": self.camera_clip_start,
            "camera_clip_end": self.camera_clip_end,
            "locked_camera_movement": self.locked_camera_movement,
            "locked_camera_rotation": self.locked_camera_rotation,
            "camera_width": self.camera_width,
            "camera_height": self.camera_height,
            "camera_pixel_size": self.camera_pixel_size,
        }

File: lib_js_naive/camera.js
/**
 * Camera - Camera configuration
 * Matches Python lib_python/camera.py
 */
class Camera {
    constructor(options = {}) {
        this.position = options.position || new Vec3(0, 0, 0);
        this.rotation = options.rotation || new Vec3(0, 0, 0);
        this.focal = options.focal || 70;
        this.clipStart = options.clipStart || 0.1;
        this.clipEnd = options.clipEnd || 100;
        this.width = options.width || 64;
        this.height = options.height || 64;
        this.pixelSize = options.pixelSize || 0.4;
    }

    clone() {
        return new Camera({
            position: this.position.clone(),
            rotation: this.rotation.clone(),
            focal: this.focal,
            clipStart: this.clipStart,
            clipEnd: this.clipEnd,
            width: this.width,
            height: this.height,
            pixelSize: this.pixelSize,
        });
    }
}

/**
 * Parse camera from JSON
 * @param {object} json - JSON object
 * @returns {Camera}
 */
function parseCamera(json) {
    return new Camera({
        position: parseVec3(json.camera_position || [0, 0, 0]),
        rotation: parseVec3(json.camera_rotation || [0, 0, 0]),
        focal: json.camera_focal || 70,
        clipStart: json.camera_clip_start || 0.1,
        clipEnd: json.camera_clip_end || 100,
        width: json.camera_width || 64,
        height: json.camera_height || 64,
        pixelSize: json.camera_pixel_size || 0.4,
    });
}

// Expose to window
window.Camera = Camera;
window.parseCamera = parseCamera;
